File: core/src/com/spacefist/entities/SpaceBlock.java
package com.spacefist.entities;

import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.math.Rectangle;
import com.badlogic.gdx.math.Vector2;
import com.spacefist.GameData;
import com.spacefist.components.NullInputComponent;
import com.spacefist.components.Physics;
import com.spacefist.components.Sound;
import com.spacefist.components.Sprite;
import org.jetbrains.annotations.NotNull;

/**
 * Represents a floating block or crate.
 */
public class SpaceBlock extends Entity {
    // Dimensions of the block
    private static final int WIDTH  = 80;
    private static final int HEIGHT = 60;

    // The sound made when the ship hits a block
    private Sound thump;
    /**
     * Creates a new SpaceBlock instance with a specified position and velocity.
     *
     * @param gameData Common game data
     * @param texture The image for the block
     * @param position The position of the block in the game world
     * @param velocity The velocity of the block
     */
    public SpaceBlock(@NotNull GameData gameData, Texture texture, @NotNull Vector2 position, Vector2 velocity) {
        super(gameData,
            new Rectangle(
                (int) position.x,
                (int) position.y,
                WIDTH,
                HEIGHT
            ),
            new Physics(),
            new NullInputComponent(),
            new Sprite(texture),
            new Sound(gameData.getSoundEffects().get("Explosion")),
            gameData.getScreenScale()
        );

        thump = new Sound(gameData.getSoundEffects().get("Explosion"));

        setVelocity(velocity);
    }

    /**
     * Called when the block has been shot
     */
    public void destroy() {
        Sound sound = (Sound) getSound();

        sound.play();
        setAlive(false);
    }

    /**
     * Called when the block has been bumped by the ship
    */
    public void thump() {
        thump.play();
        setAlive(false);
    }
}

File: core/src/com/spacefist/entities/SpaceBlock.kt
package com.spacefist.entities

import com.badlogic.gdx.graphics.Texture
import com.badlogic.gdx.math.Rectangle
import com.badlogic.gdx.math.Vector2
import com.spacefist.GameData
import com.spacefist.components.NullInputComponent
import com.spacefist.components.Physics
import com.spacefist.components.Sound
import com.spacefist.components.Sprite

/**
 * Represents a floating block or crate.
 *
 * @param gameData Common game data
 * @param texture  The image for the block
 * @param position The position of the block in the game world
 * @param velocity The velocity of the block
 */
class SpaceBlock(
    gameData: GameData,
    texture:  Texture,
    position: Vector2,
    velocity: Vector2
) : Entity(
    gameData,
    Rectangle(
        position.x.toInt().toFloat(),
        position.y.toInt().toFloat(),
        gameData.textures["Block"]!!.width.toFloat(),
        gameData.textures["Block"]!!.height.toFloat()
    ),
    Physics(),
    NullInputComponent(),
    Sprite(texture),
    Sound(gameData.soundEffects["Explosion"]!!),
    gameData.screenScale
) {

    // The sound made when the ship hits a block
    private val thump: Sound

    init {
        thump = Sound(gameData.soundEffects["Explosion"]!!)

        this.velocity = velocity
    }

    /**
     * Called when the block has been shot
     */
    fun destroy() {
        val sound = sound as Sound?

        sound!!.play()
        isAlive = false
    }

    /**
     * Called when the block has been bumped by the ship
     */
    fun thump() {
        thump.play()
        isAlive = false
    }

    companion object {
        // Dimensions of the block
        private val WIDTH  = 80
        private val HEIGHT = 60
    }
}
